### DISCUSSION TITLE
What is the correct way to use the Version Counter in an Update?

### DISCUSSION THREAD

**Discussion by benpreisler** [sqlalchemy/sqlalchemy#6607] [Usage Questions] [answered]
I'm missing something with how the the `version_id_col` version counter is supposed to work:
https://docs.sqlalchemy.org/en/14/orm/versioning.html#simple-version-counting

It's incrementing ok as long as the version is perfectly in sync with the expected version.  If i have version = 1 in the DB and I'm updating with version = 1 then it updates to 2 and everything looks good.  However if the version on the subsequent call is set to the stale version (if it was updated by another user for example) then the update logic checks against the version pulled from the DB (2) and but updates the record with the stale version (1) and throws no exception because the check 2 == 2 was valid - I'm expecting the check to be 1 == 2 and fail

```
2021-06-08 18:24:48,838 INFO sqlalchemy.engine.Engine UPDATE ex.example SET ol_version=%(ol_version)s, example_field=%(example_field)s WHERE ex.example.example_guid = %(ex_example_example_guid)s AND ex.example.ol_version = %(ex_example_ol_version)s
2021-06-08 18:24:48,838 INFO sqlalchemy.engine.Engine [cached since 0.01263s ago] {'ol_version': 1, 'example_field': 'new ex 2', 'ex_example_example_guid': 'EX-1', 'ex_example_ol_version': 2}
```

Here's my test code:
```python
from dataclasses import dataclass
from dataclasses import field

import pytest
from sqlalchemy import Column, VARCHAR, create_engine, select
from sqlalchemy.dialects.postgresql import INTEGER
from sqlalchemy.orm import registry, sessionmaker
from sqlalchemy.sql.expression import Select

from tests.testing_utils.example_entity import ExampleEntity

mapper_registry = registry()
engine = create_engine('postgresql://postgres:expwd@127.0.0.1:5111/ex_db', echo=True, future=True)
Session = sessionmaker(engine, expire_on_commit=False)


@dataclass
class BaseEntity:
    __sa_dataclass_metadata_key__ = "sa"
    __abstract__ = True
    ol_version: int = field(metadata={"sa": Column(INTEGER)}, default=None)
    __mapper_args__ = {"version_id_col": ol_version.metadata['sa']}


@mapper_registry.mapped
@dataclass
class ExampleEntity(BaseEntity):
    __tablename__ = "example"
    __table_args__ = {'schema': 'ex'}
    example_guid: str = field(metadata={"sa": Column(VARCHAR(36), primary_key=True)}, default=None)
    example_field: str = field(metadata={"sa": Column(VARCHAR)}, default=None)


@pytest.fixture()
def ex_entity():
    ee: [ExampleEntity] = ExampleEntity(
        example_guid='EX-1',
        example_field='ex field',
        ol_version=1
    )

    with Session() as session:
        with session.begin():
            session.add(ee)

    yield ee

    with Session() as session:
        with session.begin():
            entity = session.get(ExampleEntity, ee.example_guid)
            session.delete(entity)


def test_version_increment(ex_entity):
    user_1_change = ExampleEntity(
        example_guid=ex_entity.example_guid,
        example_field='new ex 1',
        ol_version=1
    )
    with Session() as session:
        with session.begin():
            updated_ee_1 = update_EE(session, user_1_change)

    assert updated_ee_1.ol_version == 2

    user_2_change = ExampleEntity(
        example_guid=ex_entity.example_guid,
        example_field='new ex 2',
        ol_version=2  # change this to 1 and the counter gets off - no exception thrown
    )

    with Session() as session:
        with session.begin():
            # I'm expecting an exception here if user_2_change.ol_version == 1
            updated_ee_2 = update_EE(session, user_2_change)

    assert updated_ee_2.ol_version == 3


def get_EE(session, example_guid):
    stmt: Select = select(ExampleEntity).where(ExampleEntity.example_guid == example_guid)
    result = session.execute(stmt)
    return result.scalars().one_or_none()


def update_EE(session, ee):
    existing_ee: ExampleEntity = get_EE(session, ee.example_guid)
    existing_ee.example_field = ee.example_field
    existing_ee.ol_version = ee.ol_version
    updated_ee: ExampleEntity = get_EE(session, ee.example_guid)
    return updated_ee
```

DDL:
```sql
CREATE TABLE ex.example
(
    example_guid  VARCHAR(36) NOT NULL,
    example_field VARCHAR,
    updated_by    INT,
    ol_version    INT         NOT NULL,
    CONSTRAINT example_pk PRIMARY KEY (example_guid)
);

CREATE TABLE ex.child
(
    child_guid   VARCHAR(36) NOT NULL,
    example_guid VARCHAR(36) NOT NULL,
    child_field  VARCHAR,
    updated_by   INT,
    ol_version   INT         NOT NULL,
    CONSTRAINT child_pk PRIMARY KEY (child_guid),
    CONSTRAINT example_fk FOREIGN KEY (example_guid) REFERENCES example (example_guid)
);
```

Thanks!

**Comment by benpreisler** [comment, marked as answer]
Ok I think I'm realizing my mistaken assumption.  I'm coming from a Spring Boot REST/JPA(Hibernate) background where the `@Version` annotated field/property and ORM functionality is built around the idea that you potentially want "optimistic locking"/version checks across web requests by different users for concurrency management. (ex: user A GETs data with a version number and then makes changes based on that data and submits an update which would fail if anything modified that record (thus version mismatch).  It looks like the version control offered by `version_id_col` is in-transaction only.  I'm guessing there is no equivalent implementation in sqlAlchemy and I'll need to implement it myself.  Does that sound correct?

**Reply by zzzeek** [reply]
hi -

the version_id_counter was designed to work exactly like the feature of Hibernate and therefore that of JPA.     I don't understand what you mean by "in transaction only".   What I dont understand in your example is why you are assigning values to the version_id_counter; you should leave it alone, and you will get failures as designed if you load the record, then seek to UPDATE it at a time that the database record itself has since been incremented.

If you could modify your test program to something I can run, that would be helpful, as it currently refers to a suite of pytest fixtures that are private to your environment.

**Reply by zzzeek** [reply]
here is a demo.  if you could illustrate what you expect to happen in this simpler test case, that would help clarify.  below, the version_id_counter is incremented automatically, I don't understand why one would be assigning it to some arbitrary number because yes that would defeat the purpose.    

```python
from sqlalchemy import Column
from sqlalchemy import create_engine
from sqlalchemy import Integer
from sqlalchemy import select
from sqlalchemy import String
from sqlalchemy.orm import declarative_base
from sqlalchemy.orm import Session
Base = declarative_base()


class A(Base):
    __tablename__ = 'a'

    id = Column(Integer, primary_key=True)
    data = Column(String(50))
    version_id = Column(Integer, nullable=False)
    __mapper_args__ = {"version_id_col": version_id}


e = create_engine("sqlite://", echo=True)
Base.metadata.drop_all(e)
Base.metadata.create_all(e)


s1 = Session(e)
s1.add(A(data='value 1'))
s1.commit()

# web request one, loads A and changes it
s2 = Session(e)
a1s2 = s2.execute(select(A)).scalar_one()
a1s2.data = "value 2"

# but web request two also loads A and changes it, more quickly
# if using repeatable
a1s1 = s1.execute(select(A)).scalar_one()
a1s1.data = "value 3"
s1.commit()

# web req one's commit will fail
s2.commit()
```

**Reply by benpreisler** [reply]
oh woops - ok I cleaned up the extraneous fixtures above

**Reply by benpreisler** [reply]
The setup I was assuming was a combination of spring boot + hibernate that checked the value of the annotated version property on the incoming object against the object from the DB.  This is for being able to send the version number with the data to the client which the client then passes back and is validated against the database to verify if the user was seeing dirty data or not.   I'm assuming just base the hibernate part works the same as sqlAlchemy - ie comparing the version on the tracked object.

**Reply by zzzeek** [reply]
that is what the feature does, so I would need more understanding of what you mean by "the client", like, a web form, then the web form wraps up the version id + data and sends back to server, then server looks up in the DB w version id, is that what you mean?    There's a way to achieve that also, let me see what your script is doing.

**Reply by zzzeek** [reply]
the code is not updated right now, you can paste runnable code in a new reply below, however read my reply below I think I know what you're looking for

**Reply by zzzeek** [reply]
OK so if I do what you're doing, and just set version_id=1 when from the DB it is 2, this just forces the version id to 1 and is not looking at that new number as what it should be comparing:

```python
# assume version id is now 2
# assume someone had version id 1 out in the wild and comes back to us
s1 = Session(e)
a1 = s1.execute(select(A)).scalar_one()
a1.data = "value 3"
a1.version_id = 1

# this does what you don't want, which is
# "UPDATE a SET data='value 3', version_id=1 WHERE a.id = 1 AND a.version_id = 2"
s1.commit()

```

So the way that SQLAlchemy supports what you are asking for right up front is by using session.merge(); since that's the use case you have here, a certain record with a PK and other data, it assumes this is a record being deserialized from the outside (such as from a web request) and it *will* look at the version id:

```python
# assume version id is now 2
# assume someone had version id 1 out in the wild and comes back to us
s1 = Session(e)
a1 = A(data="value 3", id=1, version_id=1)
a1 = s1.merge(a1)

# this raises:
# sqlalchemy.orm.exc.StaleDataError: Version id '1' on merged state <A at 0x7f7e026a1bb0> does not match existing version '2'. Leave the version attribute unset when merging to update the most recent version.
s1.commit()
```

if you set version_id=2, then it commits, and bumps the version to 3, just as you requested:

    UPDATE a SET data='value 3', version_id=3 WHERE a.id = 1 AND a.version_id = 2

There is yet another way to do this, if you didn't want to use merge(), which is that you can assign the version_id incoming as the "committed" value, that is, this will raise also:


```python
# assume version id is now 2
# assume someone had version id 1 out in the wild and comes back to us
from sqlalchemy.orm.attributes import set_committed_value
s1 = Session(e)
a1 = s1.execute(select(A)).scalar_one()
a1.data = "value 3"
set_committed_value(a1, "version_id", 1)

# this raises:
# sqlalchemy.orm.exc.StaleDataError: UPDATE statement on table 'a' expected to update 1 row(s); 0 were matched.
s1.commit()

```

so assuming I understand what you're looking for, there's two different ways to achieve it.  the session.merge() approach is the more idiomatic approach if you have the primary key available.

**Reply by benpreisler** [reply]
So in your demo example: 
- We start in DB with version_id = 1
- web request one is sent in to the server - it pulls from the DB and obtains version_id = 1
- web request two is sent - pulls and also gets version_id = 1
- web request two commits to the DB and now version_id = 2
- web request one tries to update but an error is thrown

My scenario is:
(Note - the version_id is sent to the client and returned on the subsequent PUT)
- start in DB with version_id = 1
- user 1 sends a GET to fetch the version_id = 1 data
- user 2 sends a GET to fetch the version_id = 1 data
- user 2 send web request two (PUT) with version_id = 1 to the server pulls from the DB and obtains version_id = 1
- web request two commits to the DB and now version_id = 2
- user 1 sends web request one (PUT)  with version_id = 1 to the server pulls from the DB and obtains version_id = 2
- web request one tries to update - and succeeds because version_id = 2 was pulled and then submitted
  - * this is where I need an exception thrown because web request one was trying to update using version 1 data

So essentially I need:

```python
object_to_update_from_client = A(id=1, data='value 1', version_id=1)

object_pulled_from_db = A(id=1, data='value 2', version_id=2)

if object_to_update_from_client.version_id != object_pulled_from_db.version_id:
    raise StaleDataError()
```

I was originally assuming that this check would happen internally in the library based on `version_id_col` validation (if you pass in a version_id that does not match the DB record you are trying to update then exception) but it looks like I just need to handle that myself and the `version_id_col` logic will still handle the basic check if the version changed after the record was pulled in session (tracked) from the DB

**Reply by benpreisler** [reply]
I posted this before I saw your other responses just now

**Reply by benpreisler** [reply]
ok I see your examples of merge and set_committed_value now - those do sound like what I'm looking for.  Thanks!

**Comment by vacosti** [comment]
Hello guys! I don't know if I should open another issue, so if that is the case, I'm sorry.

I have a scenario exactly like @benpreisler mentioned:

- We start in DB with version_id = 1
- web request one is sent in to the server - it pulls from the DB and obtains version_id = 1
- web request two is sent - pulls and also gets version_id = 1
- web request two commits to the DB and now version_id = 2
- web request one tries to update but an error is thrown

I have set up my tables so that an exception would be raised if the above scenario occurs (notice the version field):

```python
class Idea(Base): 
    __tablename__ = 'ideas'
    id = Column(Integer, primary_key=True, index=True)
    ...
    version = Column(Integer, nullable=False)
    __mapper_args__ = {'version_id_col': version}
```

However, I am not being able to catch the exception (sqlalchemy.orm.exc.StaleDataError) inside the context manager (as per the [docs](https://docs.sqlalchemy.org/en/20/orm/session_basics.html#session-faq-whentocreate:~:text=More%20succinctly%2C%20the%20two%20contexts%20may%20be%20combined%3A)):

```python
def stake_test(self, idea_id:int, amount_to_stake:float) -> None:
        with Session(self.DB_ENGINE, future=True) as session, session.begin():
            idea: IdeaORM | None

            try:
                idea = session.query(IdeaORM).filter(IdeaORM.id == idea_id).first()

            except Exception:
                raise CE.DBError

            if not idea:
                raise CE.IdeaDoesNotExist
    
            try:
                #this statement throws an exception in the scenario idea version was modified between read and update:
                #sqlalchemy.orm.exc.StaleDataError: UPDATE statement on table 'ideas' expected to update 1 row(s); 0 were matched.
                #However, the exception is not catched and CE.DBError below is not raised as expected
                idea.total_staked = idea.total_staked + amount_to_stake

            except Exception:
                raise CE.DBError
```

Why the exception sqlalchemy.orm.exc.StaleDataError is not being catched inside the context manager? How can I catch it?

Thank you very much

**Reply by CaselIT** [reply]
> `idea.total_staked = idea.total_staked + amount_to_stake`

this line is not flushing anything to the db, so no exception are raised there. Add `session.flush()` if you want to force a flush.

Note that for your use case the usual preferred solution is to use "for update" in the query, so that the two requests are correctly serialized by the db https://docs.sqlalchemy.org/en/20/orm/queryguide/query.html#sqlalchemy.orm.Query.with_for_update

**Reply by vacosti** [reply]
@CaselIT , thank you very much for the reply!

Some things I follow but others I don't:

1) `idea.total_staked = idea.total_staked + amount_to_stake`
This line actually _throws_ an exception: sqlalchemy.orm.exc.StaleDataError , when the version field is changed between a read and update of a table. 
The point is that I cannot catch it with a try/except block to handle it properly.

2) Following your suggestion to use "for update" (I don't know if this is the right way to use the "for update") I wrote the following code (as per the example in the [docs](https://docs.sqlalchemy.org/en/20/orm/queryguide/query.html#sqlalchemy.orm.Query.with_for_update:~:text=return%20a%20new%20Query%20with%20the%20specified%20options%20for%20the%20FOR%20UPDATE%20clause.)):

```python
def stake_test(self, idea_id:int, amount_to_stake:float) -> None:
        with Session(self.DB_ENGINE, future=True) as session, session.begin():
            idea: IdeaORM | None

            try:
                idea = session.query(IdeaORM).populate_existing().with_for_update(nowait=True, of=IdeaORM).filter(IdeaORM.id == idea_id).first()

            except Exception:
                raise CE.DBError

            if not idea:
                raise CE.IdeaDoesNotExist

            try:
                #this statement throws an exception in the scenario idea version was modified between read and update:
                #sqlalchemy.orm.exc.StaleDataError: UPDATE statement on table 'ideas' expected to update 1 row(s); 0 were matched.
                #However, the exception is not catched and CE.DBError below is not raised as expected
                idea.total_staked = idea.total_staked + amount_to_stake

            except Exception:
                raise CE.DBError
```

With the above code, I now get the following exception when a second request tries to read the database while a first request has not finished updating the database:

```
Traceback (most recent call last):
  File "/home/vacosti/cs50/consensus/consensusvenv/lib/python3.10/site-packages/sqlalchemy/engine/base.py", line 1967, in _exec_single_context
    self.dialect.do_execute(
  File "/home/vacosti/cs50/consensus/consensusvenv/lib/python3.10/site-packages/sqlalchemy/engine/default.py", line 941, in do_execute
    cursor.execute(statement, parameters)
psycopg2.errors.LockNotAvailable: could not obtain lock on row in relation "ideas"


The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/home/vacosti/ideas_backend_led/src/ideas/repository/repository.py", line 562, in stake_test2
    ideaDB = session.query(IdeaORM).populate_existing().with_for_update(nowait=True, of=IdeaORM).filter(IdeaORM.id == idea_id).first()
  File "/home/vacosti/cs50/consensus/consensusvenv/lib/python3.10/site-packages/sqlalchemy/orm/query.py", line 2728, in first
    return self.limit(1)._iter().first()  # type: ignore
  File "/home/vacosti/cs50/consensus/consensusvenv/lib/python3.10/site-packages/sqlalchemy/orm/query.py", line 2827, in _iter
    result: Union[ScalarResult[_T], Result[_T]] = self.session.execute(
  File "/home/vacosti/cs50/consensus/consensusvenv/lib/python3.10/site-packages/sqlalchemy/orm/session.py", line 2362, in execute
    return self._execute_internal(
  File "/home/vacosti/cs50/consensus/consensusvenv/lib/python3.10/site-packages/sqlalchemy/orm/session.py", line 2247, in _execute_internal
    result: Result[Any] = compile_state_cls.orm_execute_statement(
  File "/home/vacosti/cs50/consensus/consensusvenv/lib/python3.10/site-packages/sqlalchemy/orm/context.py", line 293, in orm_execute_statement
    result = conn.execute(
  File "/home/vacosti/cs50/consensus/consensusvenv/lib/python3.10/site-packages/sqlalchemy/engine/base.py", line 1418, in execute
    return meth(
  File "/home/vacosti/cs50/consensus/consensusvenv/lib/python3.10/site-packages/sqlalchemy/sql/elements.py", line 515, in _execute_on_connection
    return connection._execute_clauseelement(
  File "/home/vacosti/cs50/consensus/consensusvenv/lib/python3.10/site-packages/sqlalchemy/engine/base.py", line 1640, in _execute_clauseelement
    ret = self._execute_context(
  File "/home/vacosti/cs50/consensus/consensusvenv/lib/python3.10/site-packages/sqlalchemy/engine/base.py", line 1846, in _execute_context
    return self._exec_single_context(
  File "/home/vacosti/cs50/consensus/consensusvenv/lib/python3.10/site-packages/sqlalchemy/engine/base.py", line 1986, in _exec_single_context
    self._handle_dbapi_exception(
  File "/home/vacosti/cs50/consensus/consensusvenv/lib/python3.10/site-packages/sqlalchemy/engine/base.py", line 2355, in _handle_dbapi_exception
    raise sqlalchemy_exception.with_traceback(exc_info[2]) from e
  File "/home/vacosti/cs50/consensus/consensusvenv/lib/python3.10/site-packages/sqlalchemy/engine/base.py", line 1967, in _exec_single_context
    self.dialect.do_execute(
  File "/home/vacosti/cs50/consensus/consensusvenv/lib/python3.10/site-packages/sqlalchemy/engine/default.py", line 941, in do_execute
    cursor.execute(statement, parameters)
sqlalchemy.exc.OperationalError: (psycopg2.errors.LockNotAvailable) could not obtain lock on row in relation "ideas"

[SQL: SELECT ideas.id AS ideas_id, ideas.uuid AS ideas_uuid, ideas.creator_id AS ideas_creator_id, ideas.name AS ideas_name, ideas.uname AS ideas_uname, ideas.description AS ideas_description, ideas.total_staked AS ideas_total_staked, ideas.color AS ideas_color, ideas.image AS ideas_image, ideas.created_at AS ideas_created_at, ideas.version AS ideas_version 
FROM ideas 
WHERE ideas.id = %(id_1)s 
 LIMIT %(param_1)s FOR UPDATE OF ideas NOWAIT]
[parameters: {'id_1': 6, 'param_1': 1}]
(Background on this error at: https://sqlalche.me/e/20/e3q8)

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/vacosti/ideas_backend_led/src/ideas/test_concurrency2.py", line 36, in <module>
    test_concurrency()
  File "/home/vacosti/ideas_backend_led/src/ideas/test_concurrency2.py", line 32, in test_concurrency
    stake_repo.stake_test2(idea.id, 10)
  File "/home/vacosti/ideas_backend_led/src/ideas/repository/repository.py", line 565, in stake_test2
    raise CE.DBError
ideas.exceptions.DBError
```
So it seems that now when the first request selects the row with "for update", it locks the row so that the `psycopg2.errors.LockNotAvailable: could not obtain lock on row in relation "ideas"` is thrown. Is that correct?

3) The strange thing is that this new exception is caught by my try/except block and the other does not. I have no idea why. Do you have any ideas?

Thank you very much!

**Reply by CaselIT** [reply]
> idea.total_staked = idea.total_staked + amount_to_stake
This line actually throws an exception: sqlalchemy.orm.exc.StaleDataError , when the version field is changed between a read and update of a table.
The point is that I cannot catch it with a try/except block to handle it properly.

the fact that you cannot catch it suggest that's not where it's raised. sqlalchemy obey the classical try catch rule, so the only reason for the observed behavior is that it's not raised there. Do you have a stacktrace that says otherwise?

Regarding for-update, if you use no-wait, then yes the db will raise. if you set no-wait to false it will instead wait for the lock to be available.

>The strange thing is that this new exception is caught by my try/except block and the other does not. I have no idea why. Do you have any ideas?

that's because it's not raised inside said try catch block as mentioned

**Reply by vacosti** [reply]
@CaselIT, thank you very very much!! Setting no-wait to False solves the issue for my use case.

About the previous exception, here is the stacktrace just in case someone has the same issue:

```
Traceback (most recent call last):
  File "/home/vacosti/ideas_backend_led/src/ideas/test_concurrency1.py", line 35, in <module>
    test_concurrency()
  File "/home/vacosti/ideas_backend_led/src/ideas/test_concurrency1.py", line 29, in test_concurrency
    stake_repo.stake_test(idea.id, 10)
  File "/home/vacosti/ideas_backend_led/src/ideas/repository/repository.py", line 533, in stake_test
    with Session(self.DB_ENGINE, future=True) as session, session.begin():
  File "/home/vacosti/cs50/consensus/consensusvenv/lib/python3.10/site-packages/sqlalchemy/engine/util.py", line 147, in __exit__
    with util.safe_reraise():
  File "/home/vacosti/cs50/consensus/consensusvenv/lib/python3.10/site-packages/sqlalchemy/util/langhelpers.py", line 146, in __exit__
    raise exc_value.with_traceback(exc_tb)
  File "/home/vacosti/cs50/consensus/consensusvenv/lib/python3.10/site-packages/sqlalchemy/engine/util.py", line 145, in __exit__
    self.commit()
  File "<string>", line 2, in commit
  File "/home/vacosti/cs50/consensus/consensusvenv/lib/python3.10/site-packages/sqlalchemy/orm/state_changes.py", line 139, in _go
    ret_value = fn(self, *arg, **kw)
  File "/home/vacosti/cs50/consensus/consensusvenv/lib/python3.10/site-packages/sqlalchemy/orm/session.py", line 1313, in commit
    self._prepare_impl()
  File "<string>", line 2, in _prepare_impl
  File "/home/vacosti/cs50/consensus/consensusvenv/lib/python3.10/site-packages/sqlalchemy/orm/state_changes.py", line 139, in _go
    ret_value = fn(self, *arg, **kw)
  File "/home/vacosti/cs50/consensus/consensusvenv/lib/python3.10/site-packages/sqlalchemy/orm/session.py", line 1288, in _prepare_impl
    self.session.flush()
  File "/home/vacosti/cs50/consensus/consensusvenv/lib/python3.10/site-packages/sqlalchemy/orm/session.py", line 4352, in flush
    self._flush(objects)
  File "/home/vacosti/cs50/consensus/consensusvenv/lib/python3.10/site-packages/sqlalchemy/orm/session.py", line 4487, in _flush
    with util.safe_reraise():
  File "/home/vacosti/cs50/consensus/consensusvenv/lib/python3.10/site-packages/sqlalchemy/util/langhelpers.py", line 146, in __exit__
    raise exc_value.with_traceback(exc_tb)
  File "/home/vacosti/cs50/consensus/consensusvenv/lib/python3.10/site-packages/sqlalchemy/orm/session.py", line 4448, in _flush
    flush_context.execute()
  File "/home/vacosti/cs50/consensus/consensusvenv/lib/python3.10/site-packages/sqlalchemy/orm/unitofwork.py", line 466, in execute
    rec.execute(self)
  File "/home/vacosti/cs50/consensus/consensusvenv/lib/python3.10/site-packages/sqlalchemy/orm/unitofwork.py", line 642, in execute
    util.preloaded.orm_persistence.save_obj(
  File "/home/vacosti/cs50/consensus/consensusvenv/lib/python3.10/site-packages/sqlalchemy/orm/persistence.py", line 85, in save_obj
    _emit_update_statements(
  File "/home/vacosti/cs50/consensus/consensusvenv/lib/python3.10/site-packages/sqlalchemy/orm/persistence.py", line 948, in _emit_update_statements
    raise orm_exc.StaleDataError(
sqlalchemy.orm.exc.StaleDataError: UPDATE statement on table 'ideas' expected to update 1 row(s); 0 were matched.
```

Thanks again!

**Reply by CaselIT** [reply]
that's raised by the commit in the begin context manager, that does a flush that raises that error
